cometblue_lite/cometblue.py

```diff
@@ -61,56 +97,61 @@ def __init__(self, address, pin):
         self._status = dict()
         self._handles = dict()
         self._model = None
-#        self.update()
 
     def connect(self):
         """Connect to thermostat and send PIN"""
+        _LOGGER.debug("Connecting to device %s", self._address)
+
         try:
             self._conn.connect(self._address)
-        except btle.BTLEException as ex:
-            _LOGGER.debug("Unable to connect to the device %s, retrying: %s", self._address, ex)
+        except btle.BTLEException as exc:
+            _LOGGER.debug("Unable to connect to the device %s, retrying.", self._address, exc_info=exc)
             try:
                 self._conn.connect(self._address)
-            except Exception as ex2:
-                _LOGGER.debug("Second connection try to %s failed: %s", self._address, ex2)
+            except btle.BTLEException as exc:
+                _LOGGER.debug("Second connection try to %s failed.", self._address, exc_info=exc)
                 raise
-        if len(self._handles) <= 0:
+
+        if len(self._handles) == 0:
+            _LOGGER.debug("Discovering characteristics %s", self._address)
             try:
-                _LOGGER.debug("discovering characteristics %s", self._address)
                 service = self._conn.getServiceByUUID(COMETBLUE_SERVICE)
                 chars = service.getCharacteristics()
                 self._handles = {str(a.uuid): a.getHandle() for a in chars}
-            except btle.BTLEException as ex:
-                _LOGGER.debug("could not discover characteristics %s: %s", self._address, ex2)
+            except btle.BTLEException as exc:
+                _LOGGER.debug("Could not discover characteristics %s", self._address, exc_info=exc)
                 raise
+
+        # authenticate with PIN and initialize static values
         try:
-            self._conn.writeCharacteristic(self._handles[PASSWORD_CHAR], struct.pack(_PIN_STRUCT_PACKING, self._pin), withResponse=True)
-        except:
-            _LOGGER.info("could not write provided pin, trying 000000 %s", self._address)
-            self._conn.writeCharacteristic(self._handles[PASSWORD_CHAR], struct.pack(_PIN_STRUCT_PACKING, 0), withResponse=True)
-            
+            data = struct.pack(_PIN_STRUCT_PACKING, self._pin)
+            self._conn.writeCharacteristic(self._handles[PASSWORD_CHAR], data, withResponse=True)
+        except btle.BTLEException as exc:
+            _LOGGER.debug("Can't set PIN for device %s. Is pin=%s correct?", self._address, self._pin,
+                          exc_info=exc)
+            raise
 
     def disconnect(self):
         """Disconnect from thermostat"""
         self._conn.disconnect()
-        self._conn = btle.Peripheral()
```

Comment:
according to the bluepy documentation "a given peripheral object cannot be re-connected once connected."
are you sure this works?

Reply:
This is running without problems on my Raspberry Pi for some time now.

Judging from the bluepy code [connect()](https://github.com/IanHarvey/bluepy/blob/master/bluepy/btle.py#L441) and [disconnect()](https://github.com/IanHarvey/bluepy/blob/master/bluepy/btle.py#L447) it is possible to re-use the object.
I think, what the docs are saying, once connected, you cannot just call `connect()` with a different target, because there are no checks, if already connected. You must always first call `disconnect()` before `connect()` can be called again.